File: frontend/src/engine/grid.ts
// Brute-force / grid-search optimiser: expand a parameter grid into individual
// backtest configs, run them all, and rank by a chosen metric.
//
// Runs entirely client-side (so it works on the static site). Large grids are
// processed in chunks with progress callbacks so the UI stays responsive.

import type { BacktestConfig, Bar, Metrics, PriceLevel, Strategy } from "./types";
import { getSession } from "./sessions";
import { runBacktest } from "./backtest";

// Wait timeout used for base-strategy configs (where it's irrelevant): 48h.
const DEFAULT_ENTRY_TIMEOUT_MIN = 2880;

export interface NumRange {
  vary: boolean;
  fixed: number;
  min: number;
  max: number;
  step: number;
}

export type LevelMode = PriceLevel["mode"];
export type RankMetric =
  | "total_r"
  | "total_pnl"
  | "return_dd"
  | "profit_factor"
  | "win_rate"
  | "expectancy";

export interface GridSpec {
  strategy: Strategy;
  sessions: string[];
  directions: ("fade" | "follow")[];
  gapWindow: NumRange;
  gapSigma: NumRange;
  entryOffsetHours: NumRange; // base strategy only
  entryTimes: string[]; // follow_filters: fixed list of entry times ("HH:MM")
  // follow_filters: when varied, sweep an entry time as hours after the session
  // open (0..24); when not varied, the fixed entryTimes list above is used.
  entryTime: NumRange;
  // follow_filters: when entryTime is swept AND this is varied, add a second swept
  // entry time (hours after open). Configs then carry a two-element entry_times
  // list — two chances per cycle, first qualifying taken. Combos = time1 × time2.
  entryTime2: NumRange;
  entryTimeout: NumRange; // follow_filters: wait timeout in hours
  timeStop: { enabled: boolean } & NumRange; // hours
  sl: { enabled: boolean; mode: LevelMode } & NumRange;
  tp: { enabled: boolean; mode: LevelMode } & NumRange;
  spread: number; // static round-trip cost in price units, applied to every config
  rankBy: RankMetric;
}

export interface GridResult {
  config: BacktestConfig;
  metrics: Metrics;
}

// Hours-of-day -> "HH:MM", snapped to the 30-min bar grid and wrapped to a day
// (9.5 -> "09:30", 25.5 -> "01:30"). Wrapping lets a "hours after open" duration
// that crosses midnight resolve to the right clock time the next day.
export function hoursToHHMM(hours: number): string {
  const m = Math.round((hours * 60) / 30) * 30;
  const hh = Math.floor(m / 60) % 24;
  const mm = m % 60;
  return `${String(hh).padStart(2, "0")}:${String(mm).padStart(2, "0")}`;
}

// "HH:MM" -> hours-of-day (e.g. "09:30" -> 9.5).
export function hhmmToHours(value: string): number {
  const [h, m] = value.split(":").map(Number);
  return h + m / 60;
}

// Expand a NumRange into its values (inclusive of max, with float tolerance).
export function rangeValues(r: NumRange): number[] {
  if (!r.vary) return [r.fixed];
  if (r.step <= 0 || r.max < r.min) return [r.min];
  const out: number[] = [];
  const eps = r.step * 1e-6;
  for (let v = r.min; v <= r.max + eps; v += r.step) {
    out.push(Number(v.toFixed(6)));
  }
  return out;
}

// Value lists for each grid axis. follow_filters varies the wait timeout (not the
// entry offset) and follows only; base varies the entry offset and fade/follow.
// The unused axis collapses to length 1 so it doesn't inflate the product.
function gridAxes(spec: GridSpec) {
  const isFollow = spec.strategy === "follow_filters";
  const toMinutes = (h: number) => Math.round((h * 60) / 30) * 30;
  const directions: ("fade" | "follow")[] = isFollow
    ? ["follow"]
    : spec.directions.length
    ? spec.directions
    : ["fade"];
  // When swept, the entry time is a duration in *hours after the session open*
  // (0..24), so it can cross midnight into the next day — the actual clock time is
  // resolved per session in expandGrid (each session opens at a different hour).
  // When not swept, the fixed entry_times list is used as-is.
  const entryDurations: number[] | null =
    isFollow && spec.entryTime.vary ? rangeValues(spec.entryTime) : null;
  // Optional second swept entry time (only when the first is being swept).
  const entryDurations2: number[] | null =
    entryDurations && spec.entryTime2.vary ? rangeValues(spec.entryTime2) : null;
  return {
    isFollow,
    directions,
    entryDurations,
    entryDurations2,
    entryTimesLen: (entryDurations ? entryDurations.length : 1) * (entryDurations2 ? entryDurations2.length : 1),
    gapWindows: rangeValues(spec.gapWindow).map((v) => Math.round(v)),
    gapSigmas: rangeValues(spec.gapSigma),
    entryOffsets: isFollow ? [0] : rangeValues(spec.entryOffsetHours).map(toMinutes),
    entryTimeouts: isFollow ? rangeValues(spec.entryTimeout).map(toMinutes) : [DEFAULT_ENTRY_TIMEOUT_MIN],
    timeStops: spec.timeStop.enabled
      ? rangeValues(spec.timeStop).map(toMinutes)
      : [null as number | null],
    slValues: (spec.sl.enabled
      ? rangeValues(spec.sl).map((v) => ({ mode: spec.sl.mode, value: v }))
      : [null]) as (PriceLevel | null)[],
    tpValues: (spec.tp.enabled
      ? rangeValues(spec.tp).map((v) => ({ mode: spec.tp.mode, value: v }))
      : [null]) as (PriceLevel | null)[],
  };
}

// All backtest configs implied by the grid (the Cartesian product).
export function expandGrid(spec: GridSpec): BacktestConfig[] {
  const ax = gridAxes(spec);
  const spread = Number.isFinite(spec.spread) ? spec.spread : 0;

  const configs: BacktestConfig[] = [];
  for (const session of spec.sessions) {
    // Resolve the entry-time axis for this session: swept durations become clock
    // times anchored to *this* session's open; otherwise use the fixed list. With
    // a second swept time, each config carries both (ordered by time of day).
    const openHours = hhmmToHours(getSession(session).open_time);
    const toClock = (h: number) => hoursToHHMM(openHours + h);
    let entryTimesAxis: string[][];
    if (ax.entryDurations && ax.entryDurations2) {
      entryTimesAxis = [];
      for (const h1 of ax.entryDurations) {
        for (const h2 of ax.entryDurations2) {
          entryTimesAxis.push(h1 <= h2 ? [toClock(h1), toClock(h2)] : [toClock(h2), toClock(h1)]);
        }
      }
    } else if (ax.entryDurations) {
      entryTimesAxis = ax.entryDurations.map((h) => [toClock(h)]);
    } else {
      entryTimesAxis = [ax.isFollow ? spec.entryTimes : []];
    }
    for (const direction of ax.directions) {
      for (const gap_window of ax.gapWindows) {
        for (const gap_sigma of ax.gapSigmas) {
          for (const entry_offset_minutes of ax.entryOffsets) {
            for (const entry_times of entryTimesAxis) {
              for (const entry_timeout_minutes of ax.entryTimeouts) {
                for (const time_stop_minutes of ax.timeStops) {
                  for (const stop_loss of ax.slValues) {
                    for (const take_profit of ax.tpValues) {
                      configs.push({
                        strategy: spec.strategy,
                        session,
                        gap_window,
                        gap_sigma,
                        direction,
                        entry_offset_minutes,
                        entry_times,
                        entry_timeout_minutes,
                        adr_window: 20,
                        stop_loss,
                        take_profit,
                        time_stop_minutes,
                        intrabar: "stop_first",
                        spread,
                      });
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  return configs;
}

export function countGrid(spec: GridSpec): number {
  const ax = gridAxes(spec);
  return [
    spec.sessions.length || 1,
    ax.directions.length,
    ax.gapWindows.length,
    ax.gapSigmas.length,
    ax.entryOffsets.length,
    ax.entryTimesLen,
    ax.entryTimeouts.length,
    ax.timeStops.length,
    ax.slValues.length,
    ax.tpValues.length,
  ].reduce((a, b) => a * b, 1);
}

export function metricValue(m: Metrics, rankBy: RankMetric): number {
  switch (rankBy) {
    case "total_r":
      return m.total_r ?? -Infinity;
    case "total_pnl":
      return m.total_pnl;
    case "return_dd":
      return returnOverDrawdown(m);
    case "profit_factor":
      return m.profit_factor ?? -Infinity;
    case "win_rate":
      return m.win_rate;
    case "expectancy":
      return m.expectancy;
  }
}

// Total P/L per unit of max drawdown (a.k.a. MAR ratio) — "how much gain for the
// pain". When there's no drawdown, a profitable run ranks at the top and a
// non-profitable one at the bottom. A finite sentinel (not Infinity) keeps the
// difference-based sort well-defined.
export function returnOverDrawdown(m: Metrics): number {
  if (m.max_drawdown > 0) return m.total_pnl / m.max_drawdown;
  return m.total_pnl > 0 ? Number.MAX_VALUE : 0;
}

function configKey(c: BacktestConfig): string {
  return [
    c.strategy, c.session, c.direction, c.gap_window, c.gap_sigma, c.entry_offset_minutes,
    c.entry_times.join("/"), c.entry_timeout_minutes, c.time_stop_minutes, c.stop_loss?.mode, c.stop_loss?.value,
    c.take_profit?.mode, c.take_profit?.value,
  ].join("|");
}

// Rank best-first by the metric, with a deterministic config tiebreak so the
// result order is stable regardless of how the work was split across workers.
export function compareResults(a: GridResult, b: GridResult, rankBy: RankMetric): number {
  const d = metricValue(b.metrics, rankBy) - metricValue(a.metrics, rankBy);
  if (d !== 0) return d;
  const ka = configKey(a.config);
  const kb = configKey(b.config);
  return ka < kb ? -1 : ka > kb ? 1 : 0;
}

// Run every config, ranked best-first by the chosen metric. Chunked so the event
// loop (and progress UI) keeps ticking; `onProgress(done, total)` is called per chunk.
export async function runGrid(
  bars: Bar[],
  spec: GridSpec,
  onProgress?: (done: number, total: number) => void,
  chunkSize = 150
): Promise<GridResult[]> {
  const configs = expandGrid(spec);
  const total = configs.length;
  const results: GridResult[] = [];

  for (let i = 0; i < total; i++) {
    const config = configs[i];
    const metrics = runBacktest(bars, getSession(config.session), config).metrics;
    metrics.equity_curve = []; // not used by the optimiser report; saves memory
    results.push({ config, metrics });
    if ((i + 1) % chunkSize === 0) {
      onProgress?.(i + 1, total);
      // Yield to the event loop so the UI can repaint.
      await new Promise((r) => setTimeout(r, 0));
    }
  }
  onProgress?.(total, total);

  results.sort((a, b) => compareResults(a, b, spec.rankBy));
  return results;
}

File: backend/app/backtest/grid.py
"""Brute-force / grid-search optimiser (backend mirror of frontend/src/engine/grid.ts).

Expands a parameter grid into individual backtest configs, runs them all, and
ranks by a chosen metric. Reuses run_backtest, so per-config results match the
single-backtest path exactly.
"""
from __future__ import annotations

from itertools import product
from typing import Literal

import pandas as pd
from pydantic import BaseModel, Field

from ..sessions import DEFAULT_SESSIONS, Session
from .engine import BacktestConfig, PriceLevel, Strategy, make_runner

# Wait timeout used for base-strategy configs (where it's irrelevant): 48h.
_DEFAULT_ENTRY_TIMEOUT_MIN = 2880

LevelMode = Literal["points", "percent", "gap_multiple", "adr_multiple"]
RankMetric = Literal[
    "total_r", "total_pnl", "return_dd", "profit_factor", "win_rate", "expectancy"
]


class NumRange(BaseModel):
    vary: bool = False
    fixed: float = 0
    min: float = 0
    max: float = 0
    step: float = 1


class ToggleRange(NumRange):
    enabled: bool = False


class LevelRange(ToggleRange):
    mode: LevelMode = "adr_multiple"


class GridSpec(BaseModel):
    strategy: Strategy = Strategy.base
    sessions: list[str] = ["NY"]
    directions: list[str] = ["fade"]
    gap_window: NumRange = NumRange(fixed=20)
    gap_sigma: NumRange = NumRange(fixed=1.5)
    entry_offset_hours: NumRange = NumRange(fixed=0)  # base strategy only
    entry_times: list[str] = []  # follow_filters: fixed list of entry times ("HH:MM")
    # follow_filters: when varied, sweep an entry time as hours after the session
    # open (0..24); when not varied, entry_times is used.
    entry_time: NumRange = NumRange(fixed=9.5)
    # follow_filters: when entry_time is swept AND this is varied, add a second
    # swept entry time (hours after open) -> two-element entry_times list.
    entry_time2: NumRange = NumRange()
    entry_timeout: NumRange = NumRange(fixed=48)  # follow_filters: wait timeout in hours
    # follow_filters inversion clause: which settings to test ([False], [True], or
    # both). The multiple/offset are NumRanges so the stability sweep can vary them
    # (the optimiser keeps them fixed, vary=False).
    invert: list[bool] = [False]
    invert_multiple: NumRange = NumRange(fixed=1.0)
    invert_offset_hours: NumRange = NumRange(fixed=1.0)
    # When custom exits are on, inversion trades use invert_sl/invert_tp instead of
    # the follow trades' sl/tp.
    invert_custom_exits: bool = False
    invert_sl: LevelRange = LevelRange(enabled=False, fixed=0.5)
    invert_tp: LevelRange = LevelRange(enabled=False, fixed=1.0)
    time_stop: ToggleRange = ToggleRange(enabled=True, fixed=24)
    sl: LevelRange = LevelRange(enabled=True, fixed=0.5)
    tp: LevelRange = LevelRange(enabled=True, fixed=1.0)
    spread: float = 0.0  # static round-trip cost applied to every config
    rank_by: RankMetric = "total_r"
    top_n: int = Field(default=100, ge=1)


def _hours_to_hhmm(hours: float) -> str:
    """Hours-of-day -> "HH:MM", snapped to the 30-min grid and wrapped to a day
    (9.5 -> "09:30", 25.5 -> "01:30"). Wrapping lets a "hours after open" duration
    that crosses midnight resolve to the right clock time the next day."""
    m = int(round(hours * 60 / 30) * 30)
    return f"{(m // 60) % 24:02d}:{m % 60:02d}"


def _session_open_hours(name: str) -> float:
    """A session's open time as hours-of-day (NY 09:30 -> 9.5). Defaults to 9.5."""
    s = DEFAULT_SESSIONS.get(name)
    return (s.open_time.hour + s.open_time.minute / 60) if s else 9.5


def range_values(r: NumRange) -> list[float]:
    """Inclusive value list for a range (or just [fixed] when not varying)."""
    if not r.vary:
        return [r.fixed]
    if r.step <= 0 or r.max < r.min:
        return [r.min]
    out: list[float] = []
    eps = r.step * 1e-6
    v = r.min
    while v <= r.max + eps:
        out.append(round(v, 6))
        v += r.step
    return out


def expand_grid(spec: GridSpec) -> list[BacktestConfig]:
    # follow_filters varies the wait timeout (not the entry offset) and follows
    # only; base varies the entry offset and fade/follow. The unused axis collapses
    # to a single value so it doesn't inflate the product.
    is_follow = spec.strategy == Strategy.follow_filters
    gap_windows = [int(round(v)) for v in range_values(spec.gap_window)]
    gap_sigmas = range_values(spec.gap_sigma)
    directions = ["follow"] if is_follow else spec.directions
    entry_offsets = (
        [0] if is_follow else [int(round(h * 60 / 30) * 30) for h in range_values(spec.entry_offset_hours)]
    )
    entry_timeouts = (
        [int(round(h * 60 / 30) * 30) for h in range_values(spec.entry_timeout)]
        if is_follow
        else [_DEFAULT_ENTRY_TIMEOUT_MIN]
    )
    # When swept, the entry time is a duration in *hours after the session open*
    # (0..24), resolved to a clock time per session below (so it can cross midnight
    # into the next day). When not swept, the fixed entry_times list is used.
    swept_durations: list[float] | None = (
        range_values(spec.entry_time) if is_follow and spec.entry_time.vary else None
    )
    swept_durations2: list[float] | None = (
        range_values(spec.entry_time2) if swept_durations is not None and spec.entry_time2.vary else None
    )
    time_stops: list[int | None] = (
        [int(round(h * 60 / 30) * 30) for h in range_values(spec.time_stop)]
        if spec.time_stop.enabled
        else [None]
    )
    sls: list[PriceLevel | None] = (
        [PriceLevel(mode=spec.sl.mode, value=v) for v in range_values(spec.sl)]
        if spec.sl.enabled
        else [None]
    )
    tps: list[PriceLevel | None] = (
        [PriceLevel(mode=spec.tp.mode, value=v) for v in range_values(spec.tp)]
        if spec.tp.enabled
        else [None]
    )
    inv_custom = is_follow and spec.invert_custom_exits
    inv_sls: list[PriceLevel | None] = (
        [PriceLevel(mode=spec.invert_sl.mode, value=v) for v in range_values(spec.invert_sl)]
        if inv_custom and spec.invert_sl.enabled
        else [None]
    )
    inv_tps: list[PriceLevel | None] = (
        [PriceLevel(mode=spec.invert_tp.mode, value=v) for v in range_values(spec.invert_tp)]
        if inv_custom and spec.invert_tp.enabled
        else [None]
    )
    inverts = spec.invert if is_follow and spec.invert else [False]
    invert_multiples = range_values(spec.invert_multiple) if is_follow else [1.0]
    invert_offsets = (
        [int(round(h * 60 / 30) * 30) for h in range_values(spec.invert_offset_hours)]
        if is_follow
        else [60]
    )

    configs: list[BacktestConfig] = []
    for session in spec.sessions:
        # Resolve the entry-time axis for this session: swept durations become clock
        # times anchored to *this* session's open; otherwise use the fixed list.
        # With a second swept time, each config carries both (ordered by time).
        if swept_durations is not None and swept_durations2 is not None:
            open_h = _session_open_hours(session)
            entry_times_axis = [
                [_hours_to_hhmm(open_h + min(h1, h2)), _hours_to_hhmm(open_h + max(h1, h2))]
                for h1 in swept_durations
                for h2 in swept_durations2
            ]
        elif swept_durations is not None:
            open_h = _session_open_hours(session)
            entry_times_axis = [[_hours_to_hhmm(open_h + h)] for h in swept_durations]
        else:
            entry_times_axis = [spec.entry_times if is_follow else []]

        for direction, gw, gs, eo, ets, et, inv, invm, invo, invsl, invtp, ts, sl, tp in product(
            directions, gap_windows, gap_sigmas,
            entry_offsets, entry_times_axis, entry_timeouts, inverts,
            invert_multiples, invert_offsets, inv_sls, inv_tps, time_stops, sls, tps,
        ):
            configs.append(
                BacktestConfig(
                    strategy=spec.strategy,
                    session=session,
                    gap_window=gw,
                    gap_sigma=gs,
                    direction=direction,
                    entry_offset_minutes=eo,
                    entry_times=ets,
                    entry_timeout_minutes=et,
                    invert_enabled=inv,
                    invert_gap_multiple=invm,
                    invert_entry_offset_minutes=invo,
                    invert_custom_exits=inv_custom,
                    invert_stop_loss=invsl,
                    invert_take_profit=invtp,
                    adr_window=20,
                    stop_loss=sl,
                    take_profit=tp,
                    time_stop_minutes=ts,
                    intrabar="stop_first",
                    spread=spec.spread,
                )
            )
    return configs


def _metric_value(metrics: dict, rank_by: RankMetric) -> float:
    if rank_by == "return_dd":
        dd = metrics["max_drawdown"]
        pnl = metrics["total_pnl"]
        if dd > 0:
            return pnl / dd
        return float("inf") if pnl > 0 else 0.0
    v = metrics.get(rank_by)
    if v is None:
        return float("-inf")
    return float(v)


def run_grid(df_utc: pd.DataFrame, sessions: dict[str, Session], spec: GridSpec) -> dict:
    """Run every config and return the count plus the top-N ranked results."""
    configs = expand_grid(spec)
    run = make_runner(df_utc, sessions)  # memoizes signal-level work across configs
    results = []
    for config in configs:
        metrics = run(config)["metrics"]
        results.append({"config": config.model_dump(), "metrics": metrics})

    results.sort(key=lambda r: _metric_value(r["metrics"], spec.rank_by), reverse=True)
    return {"count": len(configs), "results": results[: spec.top_n]}
